photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataSet.scala

```diff
@@ -212,6 +239,104 @@ object LocalDataSet {
     new SparseVector(filteredIndexBuilder.result(), filteredDataBuilder.result(), features.length)
   }
 
+  /**
+    *  Compute Ratio Confidence Interval lower bounds.
+    *
+    *  @param randomLabelAndFeatures An array of (label, feature) tuples
+    *  @param quartile The quartile score of standard normal distribution
+    *  @param globalFeatureShardStats The global population feature statistics to help compute ratio
+    */
+
+  protected[ml] def computeRatioCILowerBound(
+    randomLabelAndFeatures: Array[(Double, Vector[Double])],
+    quartile: Double,
+    globalFeatureShardStats: BasicStatisticalSummary): Map[Int, Double] = {
+
+    val dummyBoundForNonBinaryOrIntercept = 2.0
+    val randomFeatureFirstOrderSums = mutable.Map[Int, Double]()
+    val lowerBounds = mutable.Map[Int, Double]()
+    var randomEffectNumSamples = 0
+    var globalPopulationNumSamples = globalFeatureShardStats.count
+    val globalPopulationFirstOrderMeans = globalFeatureShardStats.mean.toArray
+    val globallFeatureNonZero = globalFeatureShardStats.numNonzeros.toArray
+    val globalMean = globalFeatureShardStats.mean.toArray
+    randomLabelAndFeatures.foreach { case (label, features) =>
+      randomEffectNumSamples += 1
+     features.activeIterator.foreach { case (key, value) =>
+       randomFeatureFirstOrderSums.update(key, randomFeatureFirstOrderSums.getOrElse(key, 0.0) + value)}
+    }
```

Comment:
It would be faster / cleaner to do a vector sum reduce here, instead of manually iterating through the vector, e.g.:

```scala
  val randomFeatureFirstOrderSums = randomLabelAndFeatures
    .map(_._2)
    .reduce(_ + _)
```
Also, generally speaking mutability is frowned upon in the Scala world.